_posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc

```diff
@@ -0,0 +1,229 @@
+
+The Envelope format consists of the following fields:
+
+. `op` - type of operation: `c` for create, `u` for update, `d` for delete and `r` for read
+. `before` - state of row before an event occured
+. `after` - state of row after an event occurred
+. `source` - metadata of the event
+. `ts_ms` - time the connector processed the event
+
+Mapping Scylla operations to the `op` field was fairly easy: `c` for INSERT, `u` for UPDATE, `d` for DELETE.
+
+We decided to skip DELETE events that span multiple rows, such as range DELETEs: 
+
+[source,sql]
+----
+DELETE FROM ks.table 
+    WHERE pk = 1 AND ck > 0 AND ck < 5 
+----
+
+Representing such operations would unnecessarily complicate the format in order to accommodate additional range information. Moreover, it would break the expectation that an Envelope represents a modification to a single row.
+
+By default, Scylla’s CDC stores the primary key and only the modified columns of an operation. For example, suppose I created a table and inserted a row:
+
+[source,sql]
+----
+CREATE TABLE ks.example(
+    pk int, 
+    v1 int, 
+    v2 int, 
+    v3 int, 
+PRIMARY KEY(pk)) WITH cdc = {'enabled': true};
+
+INSERT INTO ks.example(pk, v1, v2, v3) 
+    VALUES (1, 2, 3, 4);
+----
+
+In Scylla you can issue another INSERT statement, which will override some of the columns:
+
+[source,sql]
+----
+INSERT INTO ks.example(pk, v1, v3) 
+    VALUES (1, 20, null);
+----
+
+The `v2` column is left unchanged after this query and we don’t have any information about its previous value. 
+
+We must be able to represent three possibilities: a column was not modified, a column was assigned a `NULL` value or a column was assigned a non-null value. The representation we chose was inspired by https://debezium.io/documentation/reference/1.6/connectors/cassandra.html[Debezium Connector for Cassandra], which works by wrapping the value for a column in a structure:
+
+[source,json]
+----
+"v1": {"value": 1},
+"v2": null,
+"v3": {"value": null}
+----
+
+A `null` structure value represents that a column was not modified (`v2` field). If the column was assigned a `NULL` value (`v3` field), there will be a structure with a `NULL` `value` field. A non-null column assignment (`v1` field) fills the contents of the `value` field. Such a format allows us to correctly represent all the possibilities and differentiate between assigning `NULL` and non-modification. 
+
+However, most sink connectors won’t be able to correctly parse such a structure. Therefore, we decided to develop our own SMT, based on Debezium’s https://debezium.io/documentation/reference/1.6/transformations/event-flattening.html[New Record State Extraction SMT]. Our https://github.com/scylladb/scylla-cdc-source-connector#scyllaextractnewstate-transformer[ScyllaExtractNewState] SMT works by applying Debezium's New Record State Extraction and flattening the `{"value": ...}` structures:
+
+[source,json]
+----
+"v1": 1,
+"v2": null,
+"v3": null
+----
```

Comment:
Isn't this exposing the same issue you seeked to circumvent with the nested representation? I.e. how will a consumer distinguish a field not modified from one whose value was set to null?

Reply:
Yes, this transformation is unfortunately lossy.